### DISCUSSION TITLE
'match' object has no attribute 'groups'

### DISCUSSION THREAD

**Discussion by prabhu-yu** [micropython/micropython#9416] [General]
In Unix port of microPython, I get the following error  
in python re module.  
'match' object has no attribute 'groups'  

However, same issue is not observed in Pico board.
I am using unix port on Ubuntu, built the code directly
using following commands.
```
cd # to your home directory
sudo apt install -y build-essential pkg-config libffi-dev
git clone https://github.com/micropython/micropython.git --branch=master --depth=1  
cd micropython  
cd ports/unix  
make submodules
make deplibs
make
make test # Optional; to test local micro python.
./build-standard/micropython # you have your microPython here.

```

**Comment by scruss** [comment]
[match.groups()](https://docs.micropython.org/en/latest/library/re.html#re.match.groups) has a note: *Note: availability of these methods depends on [MicroPython port](https://docs.micropython.org/en/latest/reference/glossary.html#term-MicroPython-port).*

I can confirm that it's missing from the unix port, and that the RP2 port does what I'd expect

**Comment by jimmo** [comment]
See https://github.com/micropython/micropython/blob/master/py/mpconfig.h#L1522 -- these `re` features are only enabled in the "Everything" configuration.

@scruss I'm not sure how it's working on rp2 which uses the "Extra features" level.

@prabhu-yu If you want to use this on Unix, you can set the options in `variants/standard/mpconfigvariant.h`.

**Reply by scruss** [reply]
@jimmo - me neither, but:
```
MicroPython v1.19.1 on 2022-09-08; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> import re
>>> m=re.match("^v[a-z][0-9]", "va3pid")
>>> m.groups()
()
>>> m.group(0)
'va3'
```

**Reply by jimmo** [reply]
Ah somehow I missed earlier that the rp2 port enables this explicitly in mpconfigport.h. I don't know why... only rp2 and mimxrt do. I suspect because they both have lots of flash.

I suspect the next step in the simplification of mpconfig features, we should take all the things that are enabled on esp32, rp2, mimxrt and use them to guide the "Full Features" level (which is currently unused) -- rp2 & unix are explicitly called out at the eventual target for the "full features" level.

**Reply by prabhu-yu** [reply]
Thank you [@jimmo](https://github.com/jimmo) .
I will try as you suggested.
I think, Unix port is the fastest way to develop/test/automate the HW independent code.
This avoids, loading the code to RP2 board for every change.
So far, I am doing it using standard Python but have to be conscious to use micro python lang. features only.
My objective is to replace above Std Python by exact microPython for the lesser surprises.
I have my own machine.py for unix which simulates uart/i2c/pin with pre defined responses
but with interface exactly as in machine.py. Thus testing app-hw interface too to certain degree w/o hw.
Yes, your thoughts on simplifying the mpconfig features will help users like me.
Thank you again.

**Comment by peterhinch** [comment]
The issue of variation between platforms was discussed extensively [here](https://github.com/micropython/micropython/issues/5564). Writing portable code is still challenging.

**Reply by prabhu-yu** [reply]
Thank you @peterhinch  for the discussion link about portable code.

**Comment by prabhu-yu** [comment]
Hi all,
Based on above discussion and help,
I created following PR.
https://github.com/micropython/micropython/pull/9513

If you have better suggestion to make std unix port similar to rp2 port,
please suggest.

Idea behind this change is to enable people to develop/test code as much as possible
on unix machine instead of going to hw each time. As said above, I have my own tiny
machine.py that enabled me to this end.
